docs/server-configuration.md
# Server Configuration

This guide will walk you through the configuration options within your `config.yaml` file.

## Prerequisites

If you haven't already, make sure you've completed the steps in the [Getting Started](getting-started.md) guide.

## 1. User configuration

Set your `name` and `client_token`. `name` identifies who you are within the Owl project. `client_token` is an authorization token used throughout the Owl project. These values are inserted automatically during the `owl setup` process.

```yaml
user:
  name: Owl
  client_token: qkirJuCHF9lKK1DvwZ2LmYSPTTFe9rR5mmJU6kp5o2M
```

> **Token Generation**:
>
> Easily generate a 32-byte (256-bit) cryptographically secure random string with this command:
>
> ```bash
> python -c 'import secrets; print(secrets.token_urlsafe(32))'
> ```

## 2. Web server configuration

Set your web server's frontend and api `base_url` and `port` values. You can also specify the environment (e.g., `development`, `production`).

```yaml
web:
  base_url: http://localhost
  port: 3000
  environment: development
  api:
    base_url: http://localhost
    port: 8000
```

> **Important**
>
> These defaults are not secure and **must** be changed in a production environment.

## 3. LLM configuration

Specify the LLM model you want to use. Owl interfaces with LLMs through [LiteLLM](https://litellm.ai/), which lets you easily use either local or remote LLMs. For a full list of LLM providers, see the [LiteLLM Providers](https://litellm.ai/providers) documentation.

### Using a local LLM

To use a local LLM, follow these steps:

1. Install [Ollama](https://github.com/ollama/ollama) and download your desired local LLM.
2. Set the `llm` section's `model` value to the chosen model.

```yaml
llm:
  model: ollama/mistral:instruct
  base_url: http://localhost
  port: 11434
  api_key:
```

### Using a remote LLM

To use a remote LLM, set the `llm` section's `model` and `key` values to your desired LLM and provider.

```yaml
llm:
  model: gpt-4
  base_url: http://localhost
  port: 11434
  api_key: <your-openai-api-key>
```

## 4. Transcription configuration

To configure speech-to-text transcription, choose a provider for both asynchronous and streaming transcription. [Deepgram](https://deepgram.com/) is recommended for ease of use. Whisper is also supported but requires additional setup.

Set your `async_transcription` and `streaming_transcription` provider to either `deepgram` or `whisper`.

```yaml
async_transcription:
  provider: deepgram

streaming_transcription:
  provider: deepgram
```

### Using Deepgram

To use Deepgram, follow these steps:

1. Create a [Deepgram](https://deepgram.com/) account to obtain an API key.
2. Configure the `deepgram` section by setting an `api_key`, `model` and `language`.

```yaml
deepgram:
  api_key: <your-deepgram-api-key>
  model: nova-2
  language: en-US
```

### Using Whisper

To use Whisper, follow these steps:

1. Create a [Hugging Face](https://huggingface.co/) account to obtain an API token. Add your Hugging Face API token to `async_whisper.hf_token`.
2. Accept the terms for the [PyAnnote Segmentation Model](https://huggingface.co/pyannote/segmentation) and [PyAnnote Speaker Diarization Model](https://huggingface.co/pyannote/speaker-diarization) on Hugging Face.

- [PyAnnote Segmentation Model](https://huggingface.co/pyannote/segmentation)
- [PyAnnote Speaker Diarization Model](https://huggingface.co/pyannote/speaker-diarization)

3. Optionally, adjust configurations for the `async_whisper` and `streaming_whisper` sections.

```yaml
async_whisper:
  host: localhost
  port: 8010
  hf_token: <your-hf-token>
  device: cpu
  compute_type: int8
  batch_size: 16
  model: tiny
  verification_threshold: 0.1
  verification_model_source: speechbrain/spkrec-ecapa-voxceleb
  verification_model_directory: .models/spkrec-ecapa-voxceleb

streaming_whisper:
  host: localhost
  port: 8009
  model: small
  language: en
  silero_sensitivity: 0.4
  webrtc_sensitivity: 2
  post_speech_silence_duration: 0.5
```

## 5. Optional configuration

### Prompt configuration

Owl uses LLM prompts for link suggestions and summarization. You can modify these prompts by changing the `prompts` section. See your generated `config.yaml` or `.config/sample-config.yaml` file for the default prompt values.

```yaml
prompt:
  suggest_links_system_message: '...'
  summarization_system_message: '...'
  short_summarization_system_message: '...'
```

## 6. Run server

Start the server using the following command:

```bash
owl serve --web
```

You can now access and use your Owl server at `http://localhost:3000`.

## Next steps

Your Owl server is now configured and ready to use. From here, you can setup an Owl capture device, learn more about Owl's configuration options or explore environment variables support.

1. View [Owl capture device](clients/) setup guides
2. Read more about Owl's [configuration options](configuration-options.md)
3. Learn about Owl's support for [environment variables](environment-variables.md)


docs/server_configuration.md
# Owl - Always-on Wearable AI

[<< Home](../README.md)

## Server Configuration

The minimal steps necessary to configure the server are listed here.

### 1. Create config.yaml in Project Root Directory

Copy `sample_config.yaml` to the root of the source tree, from where the server is run, and rename
it to `config.yaml`. When running the server, this customized configuration file can be specified by
adding the `--config=config.yaml` command line option.

Next, open `config.yaml` and configure it appropriately.

### 2. Configure User

- Set your name so the AI knows who you are and choose a secret token (a string of any length) to prevent others from accessing conversations on your server. The token will also be entered into display client applications that access the server.

```
user:
  name: "Kramer"
  client_token: 123XYZ_this_is_my_secret_token
```

### 3. Configure LLM

- [LiteLLM](https://litellm.ai/) is used as an interface to LLMs for summarization, etc. The `llm` section allows local (e.g., Ollama) or remote (e.g., OpenAI) LLMs to be specified.
- OpenAI's GPT-4 is the easiest to get up and running. [Sign up for a developer account](https://platform.openai.com/), deposit some funds for usage (we recommend beginning with ~$10), and obtain an API key. Paste that into `config.yaml`.

```
llm:
  model: gpt-4-1106-preview
  api_base_url:
  api_key: sk-a1BCDEfGhI2JKlMNOPqRS3TuvwXY4ZaBCdEFghIJK5lmnO67
```

### 4. Select Providers for Speech-to-Text Transcription

- Asynchronous transcription is used once a conversation has been detected and completed to transcribe the entire conversation.
- Streaming transcription is used for real-time transcription of speech to support assistant-in-the-loop use cases. It is only used with streaming captures but not chunked uploads.
- Deepgram is recommended for getting up and running quickly but Whisper may also be used locally if your server is capable of handling it.

```
streaming_transcription:
  provider: "deepgram"

async_transcription:
  provider: "deepgram"
```

### 5. Obtain a Deepgram API Key

- Deepgram is an online transcription service that features speaker diarization and other features. It is easy to set up, inexpensive, and new accounts are provided an ample amount of credits that should last for a long time. [Sign up for an account](https://deepgram.com/), obtain an API key, and paste it into `config.yaml`.

```
deepgram:
  api_key: "a1b2cd34ef56g7hi9012j34klmn56789012o34p5"
  model: "nova-2"
  language: "en-US"
```

- The `nova-2` model is recommended. Be aware that it has a limited number of supported languages.
- Users requiring absolute control and privacy may elect to switch away from Deepgram but it is nevertheless the easiest way to get up and running initially.

### 6. Obtain a Hugging Face API Token for Whisper and Speaker Identification Models

- Register a [Hugging Face](https://huggingface.co) account and obtain a token (these are found under "Access Tokens" in "Settings"). 
- Owl uses PyAnnote for diarization. Please visit the following Hugging Face model pages and accept the terms for each:
  - [PyAnnote Segmentation Model](https://huggingface.co/pyannote/segmentation)
  - [PyAnnote Speaker Diarization Model](https://huggingface.co/pyannote/speaker-diarization)
- The Hugging Face token is needed for asynchronous Whisper transcription (if enabled) and the SpeechBrain speaker verification model. For example:

```
async_whisper:
  host: "127.0.0.1"
  port: 8010
  hf_token: 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
  device: cpu
  compute_type: int8
  batch_size: 16
  model: tiny
  verification_threshold: 0.1
  verification_model_source: speechbrain/spkrec-ecapa-voxceleb
  verification_model_savedir: pretrained_models/spkrec-ecapa-voxceleb
```

### 7. Run the Server

- Run the server as per the [setup instructions](../README.md#server-setup) with `--config=config.yaml`.

### 8. Environment Variables

Configuration file keys can be overridden with environment variables. The format is `OWL_SECTION_KEY_NAME`. For example, the Hugging Face token under the `async_whisper` section can be set in e.g. bash with: 

```
export OWL_ASYNC_WHISPER_HF_TOKEN=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
```

On Windows, the equivalent command uses `set` and is:

```
set OWL_ASYNC_WHISPER_HF_TOKEN=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
```

However, this will create a variable that only persists in the current shell. To permanently add environment variables, press the ⊞ (Windows) key to bring up the start menu then type `environment`. Click the shortcut to the environment variable dialog.

<p align="center">
<img alt="System Properties shortcut in Start menu" src="../docs/images/windows/windows_env_vars_1.png"><br>
<i>Type "environment" in the Start menu to reveal a shortcut to the System Properties dialog.</i>
</p>

This will bring up System Properties. Click the "Environment Variables..." button on the bottom.

<p align="center">
<img alt="System Properties dialog" src="../docs/images/windows/windows_env_vars_2.png"><br>
<i>The System Properties dialog. Environment variables are accessed at the bottom.</i>
</p>

To add variables under your personal account, press the top-most "New..." button. To add them for all accounts, press the buttom one (under "System variables"). Enter each variable this way and then make sure to click OK. From now on, any new Command Prompt window will have these variables set.

<p align="center">
<img alt="Environment Variables dialog" src="../docs/images/windows/windows_env_vars_3.png"><br>
<i>Environment Variables dialog, allowing account-local or system-wide variables to be set.</i>
</p>

<p align="center">
<img alt="Entering new environment variable" src="../docs/images/windows/windows_env_vars_4.png"><br>
<i>Entering an environment variable.</i>
</p>

 

[<< Home](../README.md)

owl/core/cli.py
#
# cli.py
#
# Program entry points.
#

import asyncio
import os
import secrets
import subprocess
import time
import uuid
from datetime import datetime, timezone

import click
import requests
import uvicorn
from rich.console import Console

from alembic import command
from alembic.config import Config

from ..services.conversation.transcript_summarizer import TranscriptionSummarizer
from ..services.stt.asynchronous.async_transcription_service_factory import (
    AsyncTranscriptionServiceFactory,
)
from .config import Configuration
from .utils.setup import setup_config, setup_env

####################################################################################################
# Config File Parsing
####################################################################################################

def load_config_yaml(ctx, param, value) -> Configuration:
    return Configuration.load_config_yaml(value.name)

def add_options(options):
    def _add_options(func):
        for option in reversed(options):
            func = option(func)
        return func
    return _add_options

_config_options = [
    click.option(
        "--config",
        default="config.yaml",
        help="Configuration file",
        type=click.File(mode="r"),
        callback=load_config_yaml,
    )
]


####################################################################################################
# Help (no args)
####################################################################################################

@click.group()
def cli():
    """Owl CLI tool."""
    pass


####################################################################################################
# Setup
####################################################################################################

@cli.command()
def setup() -> None:
    """
    Set up project configuration files.
    """
    console = Console()
    console.print()
    console.rule("Owl Setup")
    console.print("\nWelcome to Owl! Let's get started.", style="bold green", end="\n\n")
    user_name: str = console.input("Enter first name: ")
    client_token: str = secrets.token_urlsafe(32)
    console.print()

    setup_env(console=console, client_token=client_token)
    setup_config(
        console=console,
        data={
            "user.name": user_name,
            "user.client_token": client_token,
        }
    )

    console.print("\nOwl setup complete. 🎉", style="bold green", end="\n\n")
    console.rule()


###################################################################################################
# Transcribe File
###################################################################################################

@cli.command()
@click.argument('main_audio_filepath', type=click.Path(exists=True))
@add_options(_config_options)
@click.option('--voice_sample_filepath', type=click.Path(exists=True), help='Path to the voice sample file')
@click.option('--speaker_name', help='Name of the speaker')
def transcribe(config: Configuration, main_audio_filepath: str, voice_sample_filepath: str, speaker_name: str):
    """Transcribe an audio file."""
    console = Console()

    console.log("[bold green]Loading transcription service...")
    start_time = time.time()

    transcription_service = AsyncTranscriptionServiceFactory.get_service(config)

    end_time = time.time()
    console.log(f"[bold green]Transcription service loaded! Time taken: {end_time - start_time:.2f} seconds")

    console.log("[cyan]Transcribing audio...")

    transcription = asyncio.run(
        transcription_service.transcribe_audio(
            main_audio_filepath, voice_sample_filepath, speaker_name
        )
    )

    console.print(transcription.utterances, style="bold yellow")
    console.log(f"[bold green]Transcription complete! Time taken: {time.time() - start_time:.2f} seconds")


###################################################################################################
# Summarize File
###################################################################################################

@cli.command()
@click.argument('main_audio_filepath', type=click.Path(exists=True))
@add_options(_config_options)
@click.option('--voice_sample_filepath', type=click.Path(exists=True), help='Path to the voice sample file')
@click.option('--speaker_name', help='Name of the speaker')
def summarize(config: Configuration, main_audio_filepath: str, voice_sample_filepath: str, speaker_name: str):
    """Transcribe and summarize an audio file."""
    console = Console()

    console.log("[bold green]Loading transcription service...")
    start_time = time.time()

    transcription_service = AsyncTranscriptionServiceFactory.get_service(config)

    end_time = time.time()
    console.log(f"[bold green]Transcription service loaded! Time taken: {end_time - start_time:.2f} seconds")

    console.log("[cyan]Transcribing audio...")

    transcription = asyncio.run(
        transcription_service.transcribe_audio(
            main_audio_filepath, voice_sample_filepath, speaker_name
        )
    )

    console.print(transcription.utterances, style="bold yellow")
    console.log(f"[bold green]Transcription complete! Time taken: {time.time() - start_time:.2f} seconds")
    summarizer = TranscriptionSummarizer(config)
    summary = asyncio.run(
        summarizer.summarize(transcription)
    )

    console.print(summary, style="bold yellow")
    console.log("[bold green]Summarization complete!")


###################################################################################################
# Send Audio
###################################################################################################

@cli.command()
@add_options(_config_options)
@click.option("--file", required=True, help="Audio file to send.")
@click.option("--timestamp", help="Timestamp in YYYYmmdd-HHMMSS.fff format. If not specified, will use current time.")
@click.option("--device-type", help="Capture device type otherwise 'unknown'.")
@click.option("--host", default="localhost", help="Address to send to.")
@click.option('--port', default=8000, help="Port to use.")
def upload(config: Configuration, file: str, timestamp: datetime | None, device_type: str | None, host: str, port: int):
    # Load file
    with open(file, "rb") as fp:
        file_contents = fp.read()

    # Create valid capture UUID
    capture_uuid = uuid.uuid1().hex

    # Timestamp
    if timestamp is not None:
        try:
           timestamp = datetime.strptime(timestamp, "%Y%m%d-%H%M%S.%f")
        except:
           raise ValueError("'timestamp' string does not conform to YYYYmmdd-HHMMSS.fff format")
    else:
        timestamp = datetime.now(timezone.utc)

    # Upload
    headers = {
        "Authorization": f"Bearer {config.user.client_token}"
    }
    data = {
         "capture_uuid": capture_uuid,
         "timestamp": timestamp.strftime("%Y%m%d-%H%M%S.%f")[:-3],
         "device_type": device_type if device_type else "unknown"
    }
    files = {
        "file": (os.path.basename(file), file_contents)
    }
    response = requests.post(url=f"http://{host}:{port}/capture/upload_chunk", files=files, data=data, headers=headers)

    # If successful, request processing
    if response.status_code != 200:
        print(f"Error {response.status_code}: {response.content}")
    else:
        response = requests.post(url=f"http://{host}:{port}/capture/process_capture", data={ "capture_uuid": capture_uuid }, headers=headers)
        if response.status_code != 200:
            print(f"Error {response.status_code}: {response.content}")
        print(response.content)

####################################################################################################
# Database
####################################################################################################

@cli.command()
@add_options(_config_options)
@click.option('--message', '-m', required=True, help='Migration message')
def create_migration(config: Configuration, message: str):
    """Generate a new migration script for schema changes."""
    console = Console()
    console.log("[bold green]Generating new migration script...")

    alembic_cfg_path = "./alembic.ini"
    alembic_cfg = Config(alembic_cfg_path)

    alembic_cfg.set_main_option("sqlalchemy.url", config.database.url)

    # Generate the migration script with Alembic
    command.revision(alembic_cfg, autogenerate=True, message=message)

    console.log(f"[bold green]Migration script generated with message: '{message}'")

####################################################################################################
# Server
####################################################################################################

@cli.command()
@click.option('--host', default='localhost', help='The interface to bind to.')
@click.option('--port', default=8000, help='The port to bind to.')
@click.option('--web', is_flag=True, help='Build and start the web frontend.')
def serve(host, port, web):
    """Start the server."""
    from .. import server
    console = Console()

    try:
        # Load the configuration
        config: Configuration = Configuration.load_config_yaml("config.yaml")
    except FileNotFoundError:
        console.log("[bold red]Error: config.yaml not found.")
        console.log("Run 'owl setup' to create the necessary configuration files.")
        return

    if web:
        console.log("[bold green]Building and starting the web frontend...")
        next_project_dir = "./clients/web"
        try:
            subprocess.run(["npm", "install"], check=True, cwd=next_project_dir)
            subprocess.Popen(["npm", "run", "dev"], cwd=next_project_dir, env=os.environ)
            console.log("[bold green]Web server started successfully.")
        except subprocess.CalledProcessError as e:
            console.log("[bold red]Failed to build or start the webserver.")
            return

    console.log(f"[bold green]Starting Python server at http://{host}:{port}...")
    app = server.create_server_app(config=config)
    uvicorn.run(app, host=host, port=port, log_level="info", ws_ping_interval=None, ws_ping_timeout=None)

if __name__ == '__main__':
    cli()
